checker/mta_sts.go

```diff
@@ -0,0 +1,154 @@
+package checker
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+func filterByPrefix(records []string, prefix string) []string {
+	filtered := []string{}
+	for _, elem := range records {
+		if elem[0:len(prefix)] == prefix {
+			filtered = append(filtered, elem)
+		}
+	}
+	return filtered
+}
+
+func getKeyValuePairs(record string, lineDelimiter string,
+	pairDelimiter string) map[string]string {
+	parsed := make(map[string]string)
+	for _, line := range strings.Split(record, lineDelimiter) {
+		split := strings.Split(strings.TrimSpace(line), pairDelimiter)
+		if len(split) != 2 {
+			continue
+		}
+		key := strings.TrimSpace(split[0])
+		value := strings.TrimSpace(split[1])
+		if parsed[key] == "" {
+			parsed[key] = value
+		} else {
+			parsed[key] = parsed[key] + " " + value
+		}
+	}
+	return parsed
+}
+
+func checkMTASTSRecord(domain string) CheckResult {
+	result := CheckResult{Name: "mta-sts-txt"}
+	records, err := net.LookupTXT(fmt.Sprintf("_mta-sts.%s", domain))
+	if err != nil {
+		return result.Failure("Couldn't find MTA-STS TXT record: %v", err)
+	}
+	return validateMTASTSRecord(records, result)
+}
+
+func validateMTASTSRecord(records []string, result CheckResult) CheckResult {
+	records = filterByPrefix(records, "v=STSv1")
+	if len(records) != 1 {
+		return result.Failure("exactly 1 MTA-STS TXT record required, found %d", len(records))
+	}
+	record := getKeyValuePairs(records[0], ";", "=")
+
+	idPattern := regexp.MustCompile("^[a-zA-Z0-9]+$")
+	if !idPattern.MatchString(record["id"]) {
+		return result.Failure("invalid id %s", record["id"])
+	}
+	return result.Success()
+}
+
+func checkMTASTSPolicyFile(domain string, hostnameResults map[string]HostnameResult) CheckResult {
+	result := CheckResult{Name: "policy_file"}
+	client := &http.Client{
+		// Don't follow redirects.
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse
+		},
+	}
+	policyURL := fmt.Sprintf("https://mta-sts.%s/.well-known/mta-sts.txt", domain)
+	resp, err := client.Get(policyURL)
+	if err != nil {
+		return result.Failure("Couldn't find policy file: %v", err)
+	}
+	if resp.StatusCode != 200 {
+		return result.Failure("Couldn't get policy file: %s", resp.Status)
+	}
+	// RFC says to check media type.
+	// if resp.Header["Content-Type"] != "text/plain" {
+	// 	return result.Error("Media type must be text/plain")
+	// }
```

Comment:
why commented?

Reply:
I was a little concerned that we'd see surprising formatting in the content type header and wasn't sure if we wanted to enforce it. I read a bit more about the expected format and also made this warning. 

From the RFC:

> When fetching a policy, senders SHOULD validate that the media type
> is "text/plain" to guard against cases where web servers allow
> untrusted users to host non-text content (typically, HTML or images)
> at a user-defined path.